### PR TITLE
#5708 Fix Dynamic Deck with Options defined

### LINKED ISSUE #5708: Collection not opened - custom study
I can't use custom study. Every time I open the deck and click on a card (to see the answer) I get the following message:

"Collection not opened - An error occurred, while accessing your collection. It may have been caused by a harmless bug, or there might be a problem with it.

Try running check database, or restoring from backup by pressing "options" button. If you continue to see this message, please file a bug report."

I did the abovementioned steps, but it didn't help. What else can I try to be able to enter custom study again?

My bug report is this:

AnkiDroid Version = 2.9.1

Android Version = 8.0.0

ACRA UUID = 5d6e2fc5-d664-46b2-b69f-cbb0be3fa9e8

###### Reproduction Steps

1. Running check data base
2. Restoring all decks from backup
3. Deleting custom study deck and tried again

###### Expected Result

I hoped to be able to access custom study.

###### Actual Result

I receive the following message: 
"Collection not opened - An error occurred, while accessing your collection. It may have been caused by a harmless bug, or there might be a problem with it..."

###### Debug info

AnkiDroid Version = 2.9.1

Android Version = 8.0.0

ACRA UUID = 5d6e2fc5-d664-46b2-b69f-cbb0be3fa9e8

###### Research
*Enter an [ x ] character to confirm the points below:*

x

[  ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[  ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[  ] I have searched for similar existing issues here and on the user forum

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description
A dynamic deck with `conf` (Deck Options) defined will crash the reviewer.

A dynamic deck should not have deck options, therefore we want to remove them.

## Fixes
Fixes #5708. But I'd either like to avoid closing the issue, or raise some other issues.

* We can improve the exception thrown in the reviewer
* We should try to track down how this happened

## Approach

This adds a task to "Check Database" to remove Deck Options from any dynamic decks.

## How Has This Been Tested?

1. Force Sync + Download corrupt collection
2. Select corrupt custom study deck and flip card (FAIL).
3. Select Check Database
4. Select same custom study deck and flip card (PASS)

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 1]
```diff
@@ -1714,6 +1715,25 @@ public long fixIntegrity(DeckTask.ProgressCallback progressCallback) {
...
+                                mDecks.removeDeckOptions(id);
+                                fixCount++;
+                            }
+                        } catch (NoSuchDeckException e) {
+                            Timber.e("Unable to find dynamic deck %d", id);
+                        }
+                    }
+                    if (fixCount > 0) {
+                        mDecks.save();
+                        problems.add(String.format(Locale.US, "%d dynamic deck(s) had deck options.", fixCount));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: Note: This is just used for logging, so it's fine to not internationalise.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 2]
```diff
@@ -1714,6 +1715,25 @@ public long fixIntegrity(DeckTask.ProgressCallback progressCallback) {
...
+                    fixIntegrityProgress(progressCallback, currentTask++, totalTasks);
+                    int fixCount = 0;
+                    for (long id : mDecks.allDynamicDeckIds()) {
+                        try {
+                            if (mDecks.hasDeckOptions(id)) {
+                                mDecks.removeDeckOptions(id);
+                                fixCount++;
+                            }
+                        } catch (NoSuchDeckException e) {
+                            Timber.e("Unable to find dynamic deck %d", id);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: This is fine to log and ignore

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/exception/NoSuchDeckException.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5186381&bid=17016315)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 3]
```diff
@@ -1714,6 +1715,25 @@ public long fixIntegrity(DeckTask.ProgressCallback progressCallback) {
...
+                        }
+                    }
+                    if (fixCount > 0) {
+                        mDecks.save();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: This doesn't touch the database, jut marks it as dirty.

**Comment by mikehardy** [comment]
If you look in AnkiDroidApp at the show startup options and dialogs you'll see a way we can force database checks when the version number passes a certain threshold - seems like we should do that in this case?

**Comment by mikehardy** [comment]
Also before you get too excited about the possibilities of refactoring that method, know that it's already in progress but has a merge/revert/needs-work status at the moment: #5172 (plus #5387 related) - but it's not a priority at the moment at all for me

**Comment by david-allison-1** [comment]
@mikehardy Agreed on the version change, but I'm unsure which version to pick, how would I go about determining it?

To me it seems like a change to `CHECK_DB_AT_VERSION`, so no refactorings needed.

I'd like to refactor a lot of the code (mostly to get rid of the linting errors and make it testable), but I don't feel it's the best use of my time while we still have major bugs.

**Comment by mikehardy** [comment]
> @mikehardy Agreed on the version change, but I'm unsure which version to pick, how would I go about determining it?
> 
> To me it seems like a change to `CHECK_DB_AT_VERSION`, so no refactorings needed.

A good question, yes it is that line, an example is this change from git blame: https://github.com/ankidroid/Anki-Android/commit/7da05d7cecb7ef518b1026f58efceba986ca67ab

You can see the versioning logic here: https://github.com/ankidroid/Anki-Android/blob/6fda6387b2f86a0653a9be24b31601fd9a6d43bc/AnkiDroid/src/main/AndroidManifest.xml#L28

The next version will be 21000150

If I pick this to v2.9.6 I will alter it post-pick to 20906201 so the 2.9.6beta1 and higher also trigger


> I'd like to refactor a lot of the code (mostly to get rid of the linting errors and make it testable), but I don't feel it's the best use of my time while we still have major bugs.

Yeah, that's where I started in 2018 and got quite a lot of test infrastructure (CI, robolectric etc) in place but then noticed that first we needed forward porting to even run on Android8+ (generating a lot of fixes for immediate needs plus some 'Deprecation' tracker issues - that's a good issue search incidentally) and that we needed some immediate crash fixes (which is the phase we're almost out of now). I think we're just about at the point where I feel like it will be possible to make larger re-factorings but that phase will start by unwinding the stack of PRs Arthur submitted, as he's already done a lot of work that had to stall for crash fixes. We're getting there though, the progress is apparent

**Comment by david-allison-1** [comment]
Rebased and done, thanks!

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/exception/NoSuchDeckException.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5186381&bid=17016315)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 4]
```diff
@@ -1606,7 +1607,7 @@ public long fixIntegrity(DeckTask.ProgressCallback progressCallback) {
         ArrayList<String> problems = new ArrayList<>();
         long oldSize = file.length();
         int currentTask = 1;
-        int totalTasks = (mModels.all().size() * 4) + 20; // 4 things are in all-models loops, 20 things are one-offs
+        int totalTasks = (mModels.all().size() * 4) + 21; // 4 things are in all-models loops, 20 things are one-offs
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: nit: should remove the counts from the comment as they will skew as they have now
```suggestion
        int totalTasks = (mModels.all().size() * 4) + 21; // a few fixes are in all-models loops, the rest are one-offs
```

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java [hunk 5]
```diff
@@ -157,7 +157,7 @@
      * collections being upgraded to (or after) this version must run an integrity check as it will contain fixes that
      * all collections should have.
      */
-    public static final int CHECK_DB_AT_VERSION = 20900148;
+    public static final int CHECK_DB_AT_VERSION = 21000150;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java]: :man_facepalming: just did 50, next is 51 sorry
```suggestion
    public static final int CHECK_DB_AT_VERSION = 21000151;
```

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 6]
```diff
@@ -1714,6 +1715,25 @@ public long fixIntegrity(DeckTask.ProgressCallback progressCallback) {
                     problems.add("Fixed " + ids.size() + " card(s) with invalid properties.");
                     mDb.execute("update cards set odid=0, odue=0 where id in " + Utils.ids2str(ids));
                 }
+                {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: none of the other ones have their own block scope - they could maybe use a newline between each fix, but they don't have independent scopes

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 7]
```diff
@@ -1714,6 +1715,25 @@ public long fixIntegrity(DeckTask.ProgressCallback progressCallback) {
...
+                    //#5708 - a dynamic deck should not have "Deck Options"
+                    fixIntegrityProgress(progressCallback, currentTask++, totalTasks);
+                    int fixCount = 0;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: maybe dynDeckFixCount? - just a more specific name so I don't think of this as global fixes

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/exception/NoSuchDeckException.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5186381&bid=17016315)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/exception/NoSuchDeckException.java  1
- AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5186381&bid=17016315)

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 6 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: Block scope gives you the ability to define more concise names (`fixCount`). I like it in this situation, as they are independent operations which ideally shouldn't be in a large function.

Would you prefer I removed it?

**Comment by david-allison-1** [comment]
> At least this way a check database will fix it any time, but if that's not already a separate issue (I have not checked) it might be a good one - ecosystem wide even (checking desktop code in ankitects repo)

It's also a crash on Anki Desktop. I'll report on the Anki Tenderapp, seems they've disabled the GitHub issues.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 6 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java]: ok - I'll go with implementor's choice ;-)

**Comment by MuhamHS** [comment]
You guys fixed it!!!!!!! THANK YOU SO MUCH!! :D Where shall I send a basket of german chocolate?!? 
I couldn't study properly since October, I almost gave up hope!!! ;D YOU GUYS ARE AWESOME!! And I am serious about the german chocolate (Best chocolate)!!!!!!!

**Comment by david-allison-1** [comment]
@MuhamHS You're very welcome! You have absolutely no idea how much I'm smiling right now. Hearing that makes all the effort worthwhile!

RE: Chocolate. Let's talk privately.

gmail: davidallisongithub

**Comment by MuhamHS** [comment]
Great!!!!! You got a mail :D

**Comment by MuhamHS** [comment]
I hope it's the right adress, since you didn't write a normal email adress

**Comment by david-allison-1** [comment]
@MuhamHS Got the email, thanks! Spambots take your email address if you post it normally.

**Pull request merged**